File: examples/python/crud_lifecycle.py
#!/usr/bin/env python3
"""End-to-end CRUD lifecycle test for the Pescheck Python client.

Exercises the full v2 surface against the API and self-cleans where possible:

  checks      list -> retrieve one
  profiles    create -> retrieve -> patch -> (appears in list) -> delete
  screenings  create (needs a profile + candidate) -> retrieve -> list documents
  webhooks    create -> list -> delete
  divisions   list -> reuse-or-create by name -> patch

Prints a line per step and "E2E CRUD OK" on success; exits non-zero on any error.

Env:
  PESCHECK_BASE_URL      e.g. https://api-staging.pescheck.io
  PESCHECK_ACCESS_TOKEN  OAuth2 client-credentials access token
"""
import os
import sys
import uuid

import pescheck
from pescheck.api.checks_api import ChecksApi
from pescheck.api.profiles_api import ProfilesApi
from pescheck.api.screenings_api import ScreeningsApi
from pescheck.api.webhooks_api import WebhooksApi
from pescheck.api.divisions_api import DivisionsApi
from pescheck.models.v2_profile_create import V2ProfileCreate
from pescheck.models.v2_profile_check import V2ProfileCheck
from pescheck.models.patched_v2_profile_partial_update import PatchedV2ProfilePartialUpdate
from pescheck.models.v2_screening_create import V2ScreeningCreate
from pescheck.models.v2_candidate import V2Candidate
from pescheck.models.webhook import Webhook
from pescheck.models.division_write import DivisionWrite
from pescheck.models.patched_division_write import PatchedDivisionWrite

# Profile/webhook names must be unique per run: the API soft-deletes and returns
# a 500 on a duplicate name. The division has no delete endpoint, so we reuse a
# fixed-named one instead of creating a new one each run.
SUFFIX = uuid.uuid4().hex[:8]
DIVISION_NAME = "E2E CI division"


def step(msg):
    print(f"  - {msg}")


def main():
    base_url = os.environ.get("PESCHECK_BASE_URL")
    token = os.environ.get("PESCHECK_ACCESS_TOKEN")
    test_email = os.environ.get("PESCHECK_TEST_EMAIL", "noreply@pescheck.nl")
    if not base_url or not token:
        print("PESCHECK_BASE_URL and PESCHECK_ACCESS_TOKEN must be set", file=sys.stderr)
        return 2

    cfg = pescheck.Configuration(host=base_url)
    cfg.access_token = token
    client = pescheck.ApiClient(cfg)

    checks = ChecksApi(client)
    profiles = ProfilesApi(client)
    screenings = ScreeningsApi(client)
    webhooks = WebhooksApi(client)
    divisions = DivisionsApi(client)

    profile_id = None
    try:
        # --- checks ---
        check_list = checks.v2_checks_list()
        check_type = check_list[0].check_type
        checks.v2_checks_retrieve(check_type)
        step(f"checks: {len(check_list)} types, retrieved '{check_type}'")

        # --- profile: create -> retrieve -> patch ---
        created = profiles.v2_profiles_create(
            V2ProfileCreate(
                name=f"E2E test profile {SUFFIX}",
                description="created by e2e",
                checks=[V2ProfileCheck(check_type=check_type)],
            )
        )
        profile_id = created.id
        profiles.v2_profiles_retrieve(profile_id)
        profiles.v2_profiles_partial_update(
            profile_id, PatchedV2ProfilePartialUpdate(description="updated by e2e")
        )
        step(f"profile: created+retrieved+patched ({profile_id})")

        # --- screening: create -> retrieve -> documents ---
        screening = screenings.v2_screenings_create(
            V2ScreeningCreate(
                profile_id=profile_id,
                candidate=V2Candidate(
                    first_name="E2E", last_name="Tester", email=test_email
                ),
            )
        )
        screenings.v2_screenings_retrieve(screening.id)
        screenings.v2_screenings_documents_list(screening.id)
        step(f"screening: created+retrieved ({screening.id}, status={screening.status})")

        # --- webhook: create -> list -> delete ---
        hook = webhooks.create_webhook2(
            Webhook(
                name=f"E2E webhook {SUFFIX}",
                url=f"https://example.com/e2e-hook-{SUFFIX}",
                events=["screening.status_changed"],
            )
        )
        webhooks.list_webhooks2()
        # webhook id deserializes as a UUID; the delete path param wants a string.
        webhooks.delete_webhook2(str(hook.id))
        step(f"webhook: created+listed+deleted ({hook.id})")

        # --- division: list -> reuse-or-create -> patch ---
        existing = divisions.v2_organisations_divisions_list().results
        div = next((d for d in existing if d.name == DIVISION_NAME), None)
        if div is None:
            div = divisions.v2_organisations_divisions_create(
                DivisionWrite(
                    name=DIVISION_NAME, city="Amsterdam", address="Teststraat 1",
                    postal="1011AA", phone="+31200000000", contact_name="E2E",
                    contact_email="e2e@example.com", invoice_email="e2e@example.com",
                )
            )
            action = "created"
        else:
            action = "reused"
        divisions.v2_organisations_divisions_partial_update(
            div.id, PatchedDivisionWrite(city="Rotterdam")
        )
        step(f"division: {action}+patched ({div.id})")

        print("E2E CRUD OK")
        return 0
    except Exception as exc:  # noqa: BLE001
        print(f"E2E CRUD FAILED: {type(exc).__name__}: {exc}", file=sys.stderr)
        return 1
    finally:
        # best-effort cleanup of the profile we created
        if profile_id:
            try:
                profiles.v2_profiles_destroy(profile_id)
            except Exception:  # noqa: BLE001
                pass


if __name__ == "__main__":
    sys.exit(main())

File: examples/ruby/crud_lifecycle.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# End-to-end CRUD lifecycle test for the Pescheck Ruby client.
#
# Exercises the full v2 surface against the API and self-cleans where possible:
#
#   checks      list -> retrieve one
#   profiles    create -> retrieve -> patch -> (appears in list) -> delete
#   screenings  create (needs a profile + candidate) -> retrieve -> list documents
#   webhooks    create -> list -> delete
#   divisions   list -> reuse-or-create by name -> patch
#
# Prints a line per step and "E2E CRUD OK" on success; exits non-zero on any error.
#
# Required environment:
#   PESCHECK_BASE_URL      e.g. https://api-staging.pescheck.io
#   PESCHECK_ACCESS_TOKEN  OAuth2 client-credentials access token
#   PESCHECK_TEST_EMAIL    candidate email (default noreply@pescheck.nl)

require 'uri'
require 'securerandom'
require 'pescheck'

# Profile/webhook names must be unique per run: the API soft-deletes and returns
# a 500 on a duplicate name. The division has no delete endpoint, so we reuse a
# fixed-named one instead of creating a new one each run.
SUFFIX = SecureRandom.hex(4)
DIVISION_NAME = 'E2E CI division'

def step(msg)
  puts "  - #{msg}"
end

base_url = ENV['PESCHECK_BASE_URL']
token = ENV['PESCHECK_ACCESS_TOKEN']
test_email = ENV.fetch('PESCHECK_TEST_EMAIL', 'noreply@pescheck.nl')

if base_url.nil? || base_url.empty? || token.nil? || token.empty?
  warn 'PESCHECK_BASE_URL and PESCHECK_ACCESS_TOKEN must be set'
  exit 2
end

uri = URI.parse(base_url)

Pescheck.configure do |config|
  config.scheme = uri.scheme
  config.host = uri.host + (uri.port && uri.default_port != uri.port ? ":#{uri.port}" : '')
  config.base_path = uri.path
  config.access_token = token
end

checks = Pescheck::ChecksApi.new
profiles = Pescheck::ProfilesApi.new
screenings = Pescheck::ScreeningsApi.new
webhooks = Pescheck::WebhooksApi.new
divisions = Pescheck::DivisionsApi.new

profile_id = nil
begin
  # --- checks ---
  check_list = checks.v2_checks_list
  check_type = check_list[0].check_type
  checks.v2_checks_retrieve(check_type)
  step("checks: #{check_list.length} types, retrieved '#{check_type}'")

  # --- profile: create -> retrieve -> patch ---
  created = profiles.v2_profiles_create(
    Pescheck::V2ProfileCreate.new(
      name: "E2E test profile #{SUFFIX}",
      description: 'created by e2e',
      checks: [Pescheck::V2ProfileCheck.new(check_type: check_type)]
    )
  )
  profile_id = created.id
  profiles.v2_profiles_retrieve(profile_id)
  profiles.v2_profiles_partial_update(
    profile_id,
    patched_v2_profile_partial_update: Pescheck::PatchedV2ProfilePartialUpdate.new(description: 'updated by e2e')
  )
  step("profile: created+retrieved+patched (#{profile_id})")

  # --- screening: create -> retrieve -> documents ---
  screening = screenings.v2_screenings_create(
    Pescheck::V2ScreeningCreate.new(
      profile_id: profile_id,
      candidate: Pescheck::V2Candidate.new(
        first_name: 'E2E', last_name: 'Tester', email: test_email
      )
    )
  )
  screenings.v2_screenings_retrieve(screening.id)
  screenings.v2_screenings_documents_list(screening.id)
  step("screening: created+retrieved (#{screening.id}, status=#{screening.status})")

  # --- webhook: create -> list -> delete ---
  hook = webhooks.create_webhook2(
    Pescheck::Webhook.new(
      name: "E2E webhook #{SUFFIX}",
      url: 'https://example.com/e2e-hook',
      events: ['screening.status_changed']
    )
  )
  webhooks.list_webhooks2
  # webhook id deserializes as a UUID; the delete path param wants a string.
  webhooks.delete_webhook2(hook.id.to_s)
  step("webhook: created+listed+deleted (#{hook.id})")

  # --- division: list -> reuse-or-create -> patch ---
  existing = divisions.v2_organisations_divisions_list.results
  div = existing.find { |d| d.name == DIVISION_NAME }
  if div.nil?
    div = divisions.v2_organisations_divisions_create(
      Pescheck::DivisionWrite.new(
        name: DIVISION_NAME, city: 'Amsterdam', address: 'Teststraat 1',
        postal: '1011AA', phone: '+31200000000', contact_name: 'E2E',
        contact_email: 'e2e@example.com', invoice_email: 'e2e@example.com'
      )
    )
    action = 'created'
  else
    action = 'reused'
  end
  divisions.v2_organisations_divisions_partial_update(
    div.id,
    patched_division_write: Pescheck::PatchedDivisionWrite.new(city: 'Rotterdam')
  )
  step("division: #{action}+patched (#{div.id})")

  puts 'E2E CRUD OK'
rescue Pescheck::ApiError => e
  warn "E2E CRUD FAILED: Pescheck::ApiError: #{e.message}"
  exit 1
rescue StandardError => e
  warn "E2E CRUD FAILED: #{e.class}: #{e.message}"
  exit 1
ensure
  # best-effort cleanup of the profile we created
  if profile_id
    begin
      profiles.v2_profiles_destroy(profile_id)
    rescue StandardError
      # ignore cleanup failures
    end
  end
end
